### PR TITLE
#27929: Create new periodic event for pruning old info about Tor routers

### PR DESCRIPTION
**Pull request by rl1987**
https://trac.torproject.org/projects/tor/ticket/27929

### CODE REVIEW

File: src/core/mainloop/mainloop.c [hunk 1]
```diff
@@ -2220,6 +2225,27 @@ retry_dns_callback(time_t now, const or_options_t *options)
   return RETRY_DNS_INTERVAL;
 }
 
+/**
+ * Periodic callback: prune routerlist of old information about Tor network.
+ */
+static int
+prune_old_routers_callback(time_t now, const or_options_t *options)
+{
+#define ROUTERLIST_PRUNING_INTERVAL (60) // 1 minute.
```
Comment (nmathewson) [inline src/core/mainloop/mainloop.c]: We could afford to make this much longer...

File: src/core/mainloop/mainloop.c [hunk 2]
```diff
@@ -1454,6 +1457,7 @@ static periodic_event_item_t *fetch_networkstatus_event=NULL;
 static periodic_event_item_t *launch_descriptor_fetches_event=NULL;
 static periodic_event_item_t *check_dns_honesty_event=NULL;
 static periodic_event_item_t *save_state_event=NULL;
+static periodic_event_item_t *prune_old_routers_event=NULL;
```
Comment (nmathewson) [inline src/core/mainloop/mainloop.c]: Do we need this?  NAMED_CALLBACK and callbacks here are only for the callbacks that we want to turn on and off by hand.

File: src/core/mainloop/mainloop.c (hunk 1 above)
Comment (nmathewson) [reply src/core/mainloop/mainloop.c]: Like, 60 minutes?

**Comment by rl1987** [comment]
These two places are `router_reload_router_list_impl()` and `router_rebuild_store()`, both of which are related persisting routerlist into file store, and may not actually call `router_reload_router_list_impl()` depending on exact codepath. I imagine the reason why this was done from `check_descriptor_callback()` is that we want to prune routerlist at some periodic interval, without depending on secondary effect of routerlist persistence logic. That would be the motivation of having it in `check_descriptor_callback()` in the first place and having it in new periodic event now.

**Pull request merged**